[quant][gpu] Adding quantized conv operator in cudnn

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70622

Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

jerryzh168 committed:
[quant][gpu] Adding quantized conv operator in cudnn

Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

Test Plan:
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

masahi commented:
Hi @jerryzh168, what is your experience with cuDNN on int8 like? Recently I tried to support cuDNN int8 on NHWC layout, but I got an impression that cuDNN does not support int32 output. What I need is `int8 data, int8 weight -> int32 output`, but I couldn't get it working. 

I noticed that you are doing `int8 data, int8 weight -> float32 output`. I also tried it but it doesn't work either (at least on NHWC input, haven't tried NCHW).

jerryzh168 commented:
> Hi @jerryzh168, what is your experience with cuDNN on int8 like? Recently I tried to support cuDNN int8 on NHWC layout, but I got an impression that cuDNN does not support int32 output. What I need is `int8 data, int8 weight -> int32 output`, but I couldn't get it working.
> 
> I noticed that you are doing `int8 data, int8 weight -> float32 output`. I also tried it but it doesn't work either (at least on NHWC input, haven't tried NCHW).

Hi @masahi, I just started with cuDNN on int8, we are planning to support native quantized GPU backend through cuDNN this half.

About the layout support, I'm using cudnn_v8 apis, and looks like the data layout is not yet exposed in the api: https://github.com/NVIDIA/cudnn-frontend/blob/main/include/cudnn_frontend_Tensor.h#L37, so my guess is that it's using the default layout (might be NCHW).

I haven't been able to get this working yet, current it can't find any engines, still debugging and trying to find the problem.

please let me know if you have NCHW working.

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
> Hi @jerryzh168, what is your experience with cuDNN on int8 like? Recently I tried to support cuDNN int8 on NHWC layout, but I got an impression that cuDNN does not support int32 output. What I need is `int8 data, int8 weight -> int32 output`, but I couldn't get it working.
> 
> I noticed that you are doing `int8 data, int8 weight -> float32 output`. I also tried it but it doesn't work either (at least on NHWC input, haven't tried NCHW).

Hi @masahi, this PR is working now, we do need NHWC layout for input, weight and output .
for dtype, the operator dtype needs to be int32, and the output tensor dtype can be int32 or float I think

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

vkuzo commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -187,6 +187,10 @@ TORCH_LIBRARY(quantized, m) {
+
+  // quantized ops implemented in cudnn, with QuantizedCUDA dispatch
+  // TODO: use the same signature as quantized::conv2d
+  m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv2d_cudnn(Tensor act, Tensor weight, Tensor? bias, int[] stride, int[] padding, int[] dilation, int groups, float output_scale, int output_zero_point) -> Tensor"));
```
I see that in native_functions.yaml there are other functions with `cudnn` in the name, I'm just wondering about the context about exposing the fact that cudnn is used versus hiding it in the implementation?

jerryzh168 replied:
we can hide it, we plan to merge this with the existing quantized::conv2d, but that's a bigger change, which involves things that is not relevant to cudnn, so want to land this one first and refactor later

jerryzh168 replied:
see the comment above the registration

zasdfgbnm commented:
cc @eqy

eqy commented:
@jerryzh168 Would it be possible to reuse v8 methods that are planned to be introduced with V8 API Convolutions? e.g., in
#60755
https://github.com/eqy/pytorch/blob/cudnn3/aten/src/ATen/native/cudnn/Conv_v8.cpp

vkuzo commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -0,0 +1,360 @@
+    // convert output fp32 Tensor to int8 by clamping
+    // TODO: get the range based on target output dtype, for now
+    // we hardcode this to the range for int8
+    Tensor clampped = output_fp32_requant.clamp(-128, 127);
```
not sure I understand this line, how does "convert fp32 to int8 by clamping" work?

jerryzh168 replied:
oh we first clamp, the type conversion is done in L347 in `clampped.to(at::kChar)`, maybe we can have a separate Tensor to make it clearer

jerryzh168 commented:
> @jerryzh168 Would it be possible to reuse v8 methods that are planned to be introduced with V8 API Convolutions? e.g., in #60755 https://github.com/eqy/pytorch/blob/cudnn3/aten/src/ATen/native/cudnn/Conv_v8.cpp

sure, when is that PR going to land? we can refactor the implementation after that is landed I think

eqy commented:
> > @jerryzh168 Would it be possible to reuse v8 methods that are planned to be introduced with V8 API Convolutions? e.g., in #60755 https://github.com/eqy/pytorch/blob/cudnn3/aten/src/ATen/native/cudnn/Conv_v8.cpp
> 
> sure, when is that PR going to land? we can refactor the implementation after that is landed I think

I hope ASAP, will bring it up in the next sync meeting. That PR would allow v7 to be continue to be used with v8 being enabled with an optional runtime flag.

vkuzo commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3886,20 +3890,27 @@ def _make_qconv_tensors(
             W = W_scale[0] * (W_init - W_zero_point[0]).float()
             b = X_scale * W_scale[0] * b_init.float()
 
+        # set Tensors to use the correct device
```
nit: could these be moved to the constructor?

jerryzh168 replied:
yeah I think so, will do

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
Currently the int8 cudnn conv is slower than fp32 cudnn conv:

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                aten::cudnn_convolution        47.28%       1.591ms        65.05%       2.189ms     109.450us     403.000us       100.00%     403.000us      20.150us            20  
                                          ProfilerStep*        17.30%     582.000us        92.87%       3.125ms     156.250us       0.000us         0.00%     403.000us      20.150us            20  
                                       cudaLaunchKernel        11.00%     370.000us        11.00%     370.000us      18.500us       0.000us         0.00%       0.000us       0.000us            20  
                                            aten::empty         6.24%     210.000us         6.24%     210.000us       3.500us       0.000us         0.00%       0.000us       0.000us            60  
                                     aten::_convolution         4.49%     151.000us        69.54%       2.340ms     117.000us       0.000us         0.00%     403.000us      20.150us            20  
                                            aten::zeros         3.74%     126.000us         6.06%     204.000us      10.200us       0.000us         0.00%       0.000us       0.000us            20  
                                      aten::convolution         3.36%     113.000us        72.90%       2.453ms     122.650us       0.000us         0.00%     403.000us      20.150us            20  
                                           aten::conv2d         1.90%      64.000us        74.80%       2.517ms     125.850us       0.000us         0.00%     403.000us      20.150us            20  
                                        cudaEventRecord         1.55%      52.000us         1.55%      52.000us       2.600us       0.000us         0.00%       0.000us       0.000us            20  
                                  cudaDeviceSynchronize         1.07%      36.000us         1.07%      36.000us      36.000us       0.000us         0.00%       0.000us       0.000us             1  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 3.365ms
Self CUDA time total: 403.000us

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                quantized::conv2d_cudnn        19.67%       2.063ms        75.84%       7.953ms     397.650us       0.000us         0.00%     787.000us      39.350us            20  
                                          ProfilerStep*        15.20%       1.594ms        91.39%       9.584ms     479.200us       0.000us         0.00%     787.000us      39.350us            20  
                                       cudaLaunchKernel        14.48%       1.519ms        14.48%       1.519ms      12.658us       0.000us         0.00%       0.000us       0.000us           120  
                                            aten::empty        12.76%       1.338ms        12.76%       1.338ms      11.150us       0.000us         0.00%       0.000us       0.000us           120  
                                    aten::empty_strided         6.52%     684.000us         6.52%     684.000us      11.400us       0.000us         0.00%       0.000us       0.000us            60  
                                         aten::int_repr         5.69%     597.000us        13.82%       1.449ms      36.225us     201.000us        19.73%     201.000us       5.025us            40  
                aten::_make_per_tensor_quantized_tensor         4.38%     459.000us         8.93%     936.000us      46.800us     220.000us        21.59%     220.000us      11.000us            20  
                                            aten::clamp         3.69%     387.000us         5.78%     606.000us      30.300us     120.000us        11.78%     120.000us       6.000us            20  
                                            aten::fill_         2.75%     288.000us         5.17%     542.000us      27.100us     100.000us         9.81%     100.000us       5.000us            20  
                                         aten::_to_copy         2.63%     276.000us         9.49%     995.000us      49.750us       0.000us         0.00%     146.000us       7.300us            20  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 10.487ms
Self CUDA time total: 1.019ms
```

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
Getting a much better time after following Yang's suggestion:

```
parameters: batch_size: 16 in_channel: 64 out_channel: 64 kernel_size: 3 height: 256 widht: 256
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                  cudaDeviceSynchronize        88.90%      64.890ms        88.90%      64.890ms      64.890ms       0.000us         0.00%       0.000us       0.000us             1  
                                aten::cudnn_convolution         9.81%       7.161ms        10.36%       7.560ms     378.000us      59.261ms        83.27%      59.261ms       2.963ms            20  
                                          ProfilerStep*         0.37%     273.000us        10.98%       8.013ms     400.650us       0.000us         0.00%      59.261ms       2.963ms            20  
                                       cudaLaunchKernel         0.37%     270.000us         0.37%     270.000us       6.750us       0.000us         0.00%       0.000us       0.000us            40  
                                            aten::empty         0.21%     150.000us         0.21%     150.000us       2.500us       0.000us         0.00%       0.000us       0.000us            60  
                                     aten::_convolution         0.10%      75.000us        10.46%       7.635ms     381.750us       0.000us         0.00%      59.261ms       2.963ms            20  
                                      aten::convolution         0.09%      63.000us        10.55%       7.698ms     384.900us       0.000us         0.00%      59.261ms       2.963ms            20  
                                            aten::zeros         0.07%      54.000us         0.13%      92.000us       4.600us       0.000us         0.00%       0.000us       0.000us            20  
                                           aten::conv2d         0.04%      30.000us        10.59%       7.728ms     386.400us       0.000us         0.00%      59.261ms       2.963ms            20  
                                        cudaEventRecord         0.03%      23.000us         0.03%      23.000us       1.150us       0.000us         0.00%       0.000us       0.000us            20  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 72.995ms
Self CUDA time total: 71.171ms

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                  cudaDeviceSynchronize        91.25%      37.069ms        91.25%      37.069ms      37.069ms       0.000us         0.00%       0.000us       0.000us             1  
                                quantized::conv2d_cudnn         2.26%     919.000us         6.95%       2.824ms     141.200us       0.000us         0.00%      18.371ms     918.550us            20  
                                          ProfilerStep*         1.54%     624.000us         8.52%       3.462ms     173.100us       0.000us         0.00%      18.371ms     918.550us            20  
                                       cudaLaunchKernel         1.23%     498.000us         1.23%     498.000us       6.225us       0.000us         0.00%       0.000us       0.000us            80  
                                            aten::empty         0.99%     401.000us         0.99%     401.000us       3.342us       0.000us         0.00%       0.000us       0.000us           120  
                                         aten::int_repr         0.75%     304.000us         1.83%     744.000us      18.600us       2.683ms         7.41%       2.683ms      67.075us            40  
                aten::_make_per_tensor_quantized_tensor         0.49%     199.000us         1.04%     423.000us      21.150us      12.248ms        33.84%      12.248ms     612.400us            20  
                                    aten::empty_strided         0.39%     158.000us         0.39%     158.000us       3.950us       0.000us         0.00%       0.000us       0.000us            40  
                                            aten::fill_         0.31%     124.000us         0.60%     243.000us      12.150us       3.440ms         9.51%       3.440ms     172.000us            20  
                          aten::_empty_affine_quantized         0.25%     102.000us         0.25%     102.000us       5.100us       0.000us         0.00%       0.000us       0.000us            20  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 40.624ms
Self CUDA time total: 36.191ms
```

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).

jerryzh168 committed:
Update on "[quant][gpu] Adding quantized conv operator in cudnn"


Summary:
This PR is the initial PR to add eager mode quantized GPU operator support, we'll start
with convolution, following cudnn fp32 Conv code and the example cudnn frontend code
https://github.com/pytorch/pytorch/pull/51390
https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fusion_sample.cpp#L557

TODO:
1. Support bias, relu, support more parameter flexibilities
2. Use the packed_prams api

Test Plan:
```
> USE_EXPERIMENTAL_CUDNN_V8_API=1 python setup.py install
> python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

debug command:
```
CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=stdout python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn > log
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33409155](https://our.internmc.facebook.com/intern/diff/D33409155)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33409155).
